File: leetcode/rust/src/213.house-robber-ii.rs
use std::cmp::max;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.is_empty() {
            return 0;
        }

        if nums.len() == 1 {
            return nums[0];
        }


        max(Self::_rob(&nums[..nums.len() - 1]), Self::_rob(&nums[1..]))
    }

    pub fn _rob(nums: &[i32]) -> i32 {
        if nums.is_empty() {
            return 0;
        }

        if nums.len() == 1 {
            return nums[0];
        }

        let mut res: Vec<i32> = vec![nums[0], max(nums[0], nums[1])];

        for i in 2..nums.len() {
            res.push(max(nums[i] + res[i-2], res[i - 1]));
        }

        *res.last().unwrap()
    }
}

File: leetcode/kt/213.house-robber-ii.kt
import kotlin.math.max

class Solution {
    fun rob(nums: IntArray): Int {
        if(nums.isEmpty())
            return 0
        if(nums.size == 1)
            return nums.last()
        if(nums.size == 2)
            return max(nums[0],nums[1])
        
        val left = doRob(nums,0,nums.size-1);
        val right = doRob(nums,1,nums.size);

        return max(left,right)
    }

    fun doRob(nums: IntArray, from: Int, to: Int): Int {
        var prev : Int = nums[from]
        var curr = max(nums[from],nums[from+1])
        for(i in from + 2 until to){
            val now = max(prev + nums[i],curr)
            prev = curr
            curr = now;
        }
        return curr
    }
}